pkg/autoscaler/metrics/stats_scraper_test.go

```diff
@@ -152,47 +181,149 @@ func TestNewServiceScraperWithClientErrorCases(t *testing.T) {
+func TestPodDirectScrape1Failure(t *testing.T) {
+	fake.KubeClient = fakek8s.NewSimpleClientset()
+	fake.KubeInformer = kubeinformers.NewSharedInformerFactory(fake.KubeClient, 0)
+
+	// For 5 pods, we need 4 successes.
+	client := newTestScrapeClient(testStats, []error{nil, nil, errors.New("okay"), nil, nil})
+	scraper, err := serviceScraperForTest(client, true)
+	if err != nil {
+		t.Fatal("serviceScraperForTest:", err)
+	}
+	fake.Endpoints(5, fake.TestService)
+	makePods(5)
 	now := time.Now()
+
 	got, err := scraper.Scrape(defaultMetric.Spec.StableWindow)
 	if err != nil {
-		t.Fatalf("Unexpected error from scraper.Scrape(): %v", err)
+		t.Fatal("Unexpected error from scraper.Scrape():", err)
 	}
 
 	if got.Time.Before(now) {
-		t.Errorf("stat.Time=%v, want bigger than %v", got.Time, now)
+		t.Errorf("stat.Time=%v, want greater than %v", got.Time, now)
 	}
-	if got.PodName != scraperPodName {
-		t.Errorf("stat.PodName=%v, want %v", got.PodName, scraperPodName)
+	// Checking one of the metrics is enough here.
+	if got.AverageConcurrentRequests != 20.0 {
+		t.Errorf("stat.AverageConcurrentRequests=%v, want %v",
+			got.AverageConcurrentRequests, 20.0)
 	}
-	// (3.0 + 5.0 + 3.0) / 3.0 * 3 = 11
-	if got.AverageConcurrentRequests != 11.0 {
+
+	if !scraper.podsAddressable {
+		t.Error("PodAddressable switched to false")
+	}
+}
+
+func TestPodDirectScrapeAllFail(t *testing.T) {
+	fake.KubeClient = fakek8s.NewSimpleClientset()
+	fake.KubeInformer = kubeinformers.NewSharedInformerFactory(fake.KubeClient, 0)
+
+	// For 5 pods, we need 4 successes.
+	testStats := testStatsWithTime(4, youngPodCutOffDuration.Seconds() /*youngest*/)
+	client := newTestScrapeClient(testStats, []error{
+		// Pods fail.
+		errors.New("okay"), errors.New("okay"), errors.New("okay"), errors.New("okay"),
+		// Service succeeds.
+		nil, nil, nil, nil,
+	})
+	scraper, err := serviceScraperForTest(client, true)
+	if err != nil {
+		t.Fatal("serviceScraperForTest:", err)
+	}
+	fake.Endpoints(4, fake.TestService)
+	makePods(4)
+	now := time.Now()
+
+	got, err := scraper.Scrape(defaultMetric.Spec.StableWindow)
+	if err != nil {
+		t.Fatal("Unexpected error from scraper.Scrape():", err)
+	}
+
+	if got.Time.Before(now) {
+		t.Errorf("stat.Time=%v, want greater than %v", got.Time, now)
+	}
+	// Checking one of the metrics is enough here.
+	if got.AverageConcurrentRequests != 20.0 {
 		t.Errorf("stat.AverageConcurrentRequests=%v, want %v",
-			got.AverageConcurrentRequests, 11.0)
+			got.AverageConcurrentRequests, 20.0)
 	}
-	// ((5 + 7 + 5) / 3.0) * 3 = 17
-	if got.RequestCount != 17 {
-		t.Errorf("stat.RequestCount=%v, want %v", got.RequestCount, 15)
+
+	if scraper.podsAddressable {
+		t.Error("PodAddressable didn't switch to true")
 	}
-	// (2.0 + 4.0 + 2.0) / 3.0 * 3 = 8
-	if got.AverageProxiedConcurrentRequests != 8.0 {
-		t.Errorf("stat.AverageProxiedConcurrentRequests=%v, want %v",
-			got.AverageProxiedConcurrentRequests, 8.0)
+}
+
+func TestPodDirectScrapeSomeFail(t *testing.T) {
```

Comment:
The test name doesn't reflect the difference between `TestPodDirectScrape1Failure`. How about `TestPodDirectPodExausted`?

Reply:
Renamed.